### DISCUSSION TITLE
Models not loading correctly after updating Mesa Geo Package

### DISCUSSION THREAD

**Discussion by sknill-tm** [mesa/mesa-geo#126] [Q&A]
Hello, I have recently update the mesa-geo package to the latest version (believe I was going from 1.2 to 4.0) and this has thrown up a few errors with my existing code. I have made the relevant changes so the code should be operational with 4.0 but for some reason the server loads but it has not been populated. See below:

![image](https://user-images.githubusercontent.com/102371908/204766416-4218eb8e-674b-432a-bbfe-2c7f2fdca6cc.png)

In an attempt to get the code working again, I have created a separate environment to host a previous state of the mesa geo packages so that I could run my legacy code. I am able to get the server to load and the relevant agents to populate, however the model will not actually run or step. 

Appreciate the support.

**Comment by rht** [comment]
Any error messages from the browser console or the Python server itself?

**Reply by sknill-tm** [reply]
**Scenario 1**
When running the model with the latest version of mesa (1.1.1) and mesa-geo (0.4.0), running both an updated version of my code and/or the latest version of the GEOSIR example the server opens but no agents are populated and it is not possible to run the model (previous screenshot).

IPython Console message:

"
Interface starting at http://127.0.0.1:8521
Traceback (most recent call last):

  File ~\OneDrive - Techmodal\Modelling\Capability Development\ABS\Python\Ukraine Refugee Model\run.py:3
    server.launch()

  File ~\AppData\Local\Continuum\Anaconda3\envs\mgeotest\lib\site-packages\mesa\visualization\ModularVisualization.py:410 in launch
    tornado.ioloop.IOLoop.current().start()

  File ~\AppData\Local\Continuum\Anaconda3\envs\mgeotest\lib\site-packages\tornado\platform\asyncio.py:215 in start
    self.asyncio_loop.run_forever()

  File ~\AppData\Local\Continuum\Anaconda3\envs\mgeotest\lib\asyncio\base_events.py:591 in run_forever
    self._check_running()

  File ~\AppData\Local\Continuum\Anaconda3\envs\mgeotest\lib\asyncio\base_events.py:583 in _check_running
    raise RuntimeError('This event loop is already running')

RuntimeError: This event loop is already running

"

**Scenario 2**
When attempting to run the model using the previous versions of mesa (0.9.0) and mesa-geo (0.1.2) in a new environment - the server opens and populates the agents, but does not run. 

IPython Console message:

"
C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\pyproj\crs\crs.py:280: FutureWarning: '+init=<authority>:<code>' syntax is deprecated. '<authority>:<code>' is the preferred initialization method. When making the change, be mindful of axis order changes: https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6
  projstring = _prepare_from_string(projparams)

ERROR:tornado.application:Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET', uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\tornado\websocket.py", line 635, in _run_callback
    result = callback(*args, **kwargs)
  File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\mesa_geo\visualization\ModularVisualization.py", line 199, in on_message
    self.application.model.step()
  File "C:\Users\s.knill\OneDrive - Techmodal\Modelling\Capability Development\ABS\Python\Ukraine Refugee Model\model.py", line 334, in step
    self.grid._recreate_rtree(PersonAgent)  # Recalculate spatial tree, because agents are moving
  File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\mesa_geo\geospace.py", line 147, in _recreate_rtree
    agents = old_agents + new_agents
TypeError: can only concatenate list (not "type") to list

"

Thanks

**Reply by rht** [reply]
For scenario 1: do you mind creating a minimal reproducer for your model that can be posted publicly?
For scenario 2: `self.grid._recreate_rtree(PersonAgent)` looks wrong, because `PersonAgent` is a type, not a list.

**Reply by rht** [reply]
Are you running inside a Jupyter notebook? What if you run the server from the command line?

**Reply by sknill-tm** [reply]
Scenario 1: I will look in to this. 
Scenario 2: I have updated line 334 self.grid._recreate_rtree(PersonAgent) --> self.grid._recreate_rtree() - same situation map with populated agents but does not run. 

New error message below:
"
ERROR:tornado.application:Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET', uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\tornado\websocket.py", line 635, in _run_callback
    result = callback(*args, **kwargs)
  File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\mesa_geo\visualization\ModularVisualization.py", line 199, in on_message
    self.application.model.step()
  File "C:\Users\s.knill\OneDrive - Techmodal\Modelling\Capability Development\ABS\Python\Ukraine Refugee Model\model.py", line 334, in step
    self.grid._recreate_rtree()  # Recalculate spatial tree, because agents are moving
TypeError: _recreate_rtree() missing 1 required positional argument: 'new_agents'
" 

Running it in spyder.

**Reply by rht** [reply]
Can you try passing an empty list to `_recreate_rtree()`?
Can you try running for both scenarios in a command line? Either Git Bash or Powershell. I'm not sure, but it's possible that there is a Tornado instance running in the Spyder already.

**Reply by sknill-tm** [reply]
I have passed an empty list to _recreate_retree() and it now seems to work (although it seems to iterate over multiple steps before actually updating).

After the first step the below error message appears:
"
Exception ignored in: <function Handle.del at 0x000002A1BAD56940>
Traceback (most recent call last):
File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\rtree\index.py", line 1234, in del
self.destroy()
File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\rtree\index.py", line 1221, in destroy
self._destroy(self._ptr)
File "C:\Users\s.knill\AppData\Local\Continuum\Anaconda3\envs\mgeo\lib\site-packages\rtree\core.py", line 34, in check_void_done
raise RTreeError(msg)
rtree.exceptions.RTreeError: Error in "IndexProperty_Destroy": Property IndexIdentifier was empty
"
Apologies I am relatively new to this and wasn't able to get either scenario to run in Powershell. Is there a way to close the Tornado instance through spyder?

**Reply by rht** [reply]
@wang-boyu could give insight to the `rtree` stuff, but I think we should just focus on fixing on Mesa-Geo 4.0 instead. I will wait for your minimal reproducer.

Another sure way to find the commit that causes the problem is by using `git bisect`.

**Reply by sknill-tm** [reply]
Thanks for the help - will wait to see wang-boyu's input on the rtree stuff.

Yeah, it would be good to get Mesa-Geo 4.0 working and makes sense going forward. At the moment I cannot get the examples (GEOSIR or GeoSchelling) working - assume this is not an issue with the code but rather the way the packages have been installed for Mesa-Geo?

current env package versions:
------------------ ------------
affine             2.3.1
appdirs            1.4.4
arrow              1.2.3
asttokens          2.0.5
attrs              22.1.0
backcall           0.2.0
beautifulsoup4     4.11.1
binaryornot        0.4.4
branca             0.6.0
certifi            2022.9.24
chardet            5.0.0
charset-normalizer 2.1.1
click              7.1.2
click-plugins      1.1.1
cligj              0.7.2
cloudpickle        2.0.0
colorama           0.4.6
cookiecutter       2.1.1
debugpy            1.5.1
decorator          5.1.1
executing          0.8.3
Fiona              1.8.13.post1
folium             0.13.0
GDAL               3.0.2
geojson            2.5.0
geopandas          0.12.1
idna               3.4
ipykernel          6.15.2
ipython            8.6.0
jedi               0.18.1
Jinja2             3.1.2
jinja2-time        0.2.0
jupyter-client     6.1.12
jupyter_core       4.11.2
libpysal           4.7.0
MarkupSafe         2.1.1
matplotlib-inline  0.1.6
Mesa               1.1.1
Mesa-Geo           0.4.0
mkl-fft            1.3.1
mkl-random         1.2.2
mkl-service        2.4.0
munch              2.5.0
nest-asyncio       1.5.5
networkx           2.8.8
numpy              1.23.4
packaging          21.3
pandas             1.5.2
parso              0.8.3
pickleshare        0.7.5
pip                22.2.2
platformdirs       2.5.4
prompt-toolkit     3.0.20
psutil             5.9.0
pure-eval          0.2.2
Pygments           2.11.2
pyparsing          3.0.9
pyperclip          1.8.2
pyproj             2.6.1.post1
python-dateutil    2.8.2
python-slugify     7.0.0
pytz               2022.6
pywin32            302
PyYAML             6.0
pyzmq              23.2.0
rasterio           1.3.4
requests           2.28.1
Rtree              0.9.7
scipy              1.9.3
setuptools         65.5.0
Shapely            1.8.4
six                1.16.0
snuggs             1.4.7
soupsieve          2.3.2.post1
spyder-kernels     2.1.3
stack-data         0.2.0
text-unidecode     1.3
tornado            6.2
tqdm               4.64.1
traitlets          5.1.1
urllib3            1.26.13
wcwidth            0.2.5
wheel              0.37.1
wincertstore       0.2

**Reply by wang-boyu** [reply]
> TypeError: _recreate_rtree() missing 1 required positional argument: 'new_agents'

This is weird because `new_agents` is optional and defaults to `None`: https://github.com/projectmesa/mesa-geo/blob/99eeec3f5d788e98c2473a4a0e5be923800408f5/mesa_geo/geospace.py#L166

Could you please share your Python version and how Mesa-Geo was installed? If possible, could you have a separate, clean virtual environment and re-install Mesa-Geo from scratch?

And yes a minimal reproducer of the model would be great.

By the way, in [v0.3.0](https://github.com/projectmesa/mesa-geo/releases/tag/v0.3.0) there are two breaking changes:

- rename `model.grid` to `model.space` https://github.com/projectmesa/mesa-geo/pull/40
- rename GeoAgent's `shape` attribute to `geometry` https://github.com/projectmesa/mesa-geo/pull/57

Not sure whether this will help or not.

**Reply by sknill-tm** [reply]
Thanks, creating a new virtual env with Python 3.9 and Mesa-Geo 0.4 appears to have gotten the model working again (agents populate and model runs as expected). However, the agent attributes are not displaying anymore (see screenshot below) is this to do with another revision in the code?

![image](https://user-images.githubusercontent.com/102371908/205321900-3baada15-b7b8-4382-9eae-45a50b936ea8.png)

Noted re the breaking points - these have all been updated in the code. 

Thanks again for all the help!

**Reply by wang-boyu** [reply]
Excellent!

To display agent attributes, you will need to specify them in the `portrayal_method` function, with a `description` key in the returned dictionary.

For example to show agent's `unique_id` attribute:

```python
def schelling_draw(agent):
    """
    Portrayal Method for canvas
    """
    portrayal = dict()
    portrayal["description"] = agent.unique_id  # <------- add this for display
    if agent.atype is None:
        portrayal["color"] = "Grey"
    elif agent.atype == 0:
        portrayal["color"] = "Red"
    else:
        portrayal["color"] = "Blue"
    return 
```

It is documented here: https://github.com/projectmesa/mesa-geo/blob/99eeec3f5d788e98c2473a4a0e5be923800408f5/mesa_geo/visualization/modules/MapVisualization.py#L37-L40

Let me know whether this works for you.

**Reply by sknill-tm** [reply]
Thanks for sharing. 

I have tried to pull through the ID and agent type, but it looks like only the second call (agent type) is pulling through. They also seem to pull through in a vertical fashion (see below)

![image](https://user-images.githubusercontent.com/102371908/205868101-92da09e7-b8ce-45d9-970f-b1d08cd5ed3d.png)

**Reply by wang-boyu** [reply]
Sorry I might have made a mistake in previous comments. Could you try to define the `description` field itself as a dictionary?

For example:

```python
def schelling_draw(agent):
    """
    Portrayal Method for canvas
    """
    portrayal = dict()
    
    # add this for display
    portrayal["description"] = {
        "id": agent.unique_id,
        "other_property": agent.other_property,
    }

    if agent.atype is None:
        portrayal["color"] = "Grey"
    elif agent.atype == 0:
        portrayal["color"] = "Red"
    else:
        portrayal["color"] = "Blue"
    return 
```
